File: Leetcode/Easy/20. Valid Parentheses/solution.py
class Solution:
    # @return a boolean
    def isValid(self, s):
        
        if len(s) % 2 != 0:
            return False
        
        stack = []
        dict = {
            "]":"[", 
            "}":"{", 
            ")":"("
        }
        
        for char in s:
            if char in dict.values():
                stack.append(char)
            elif char in dict.keys():
                if stack == [] or dict[char] != stack.pop():
                    return False
            else:
                return False
        return len(stack) == 0

File: Leetcode/Easy/20. Valid Parentheses/solution.ts
function isValid(s: string): boolean {
    const closing = [')', ']', '}'];
    const opening = ['(', '[', '{'];
    
    let stack = [], stackLen = 0;
    
    for (let i = 0; i < s.length; i++) {
        if (closing.includes(s[i])) {
            const lastOpening = stack.pop();
            stackLen--;
            if (opening.indexOf(lastOpening) !== closing.indexOf(s[i])) {
                return false;
            }
        } else {
            stack.push(s[i]);
            stackLen++;
        }
    }
    
    if (stackLen > 0) {
        return false;
    } else {
        return true;
    }
    
};